### PR TITLE
feat(api)!: Web Socket - Reconnection

### PR DESCRIPTION
**Pull request by Equartey**
*Description of changes:*

- Adds reconnection logic to web socket refactor
- Adds new tests & updates previous ones
- Removes old web socket connection files
- Clean up around the bloc

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

### CODE REVIEW

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 1]
```diff
@@ -252,19 +316,96 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
+    );
+    final state = _currentState;
+    try {
+      // Begin reconnection with retry/back off on ping endpoint
+      final res = await state.options.retryOptions.retry(
+        // Make a GET request
+        () => _getPollRequest().timeout(_defaultRetryTimeout),
+      );
+
+      final body = await utf8.decodeStream(res.body);
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:371]: This logic is repeated in `_poll`. Can it be DRY'd up?

Comment (dnys1) [reply packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:371]:
```suggestion
      final body = await res.decodeBody();
```

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 2]
```diff
@@ -252,19 +316,96 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
+          recoverySuggestion:
+              'Unable to reach the configured AppSync URL. Check internet connection or the configured AppSync URL',
+        );
+      }
+
+      // **Ping succeeded**
+
+      // Prep new connection
+      await state.service.close();
+      await Future.wait(
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:385]: Can be a normal `for in` loop since there is no async work

File: packages/api/amplify_api/test/graphql_test.dart [hunk 3]
```diff
@@ -293,42 +301,63 @@ void main() {
...
-
-      // WebSocketConnection.httpClientOverride = mockHttpClient;
-      // WebSocketConnection.connectivityOverride = connectivity;
-      // WebSocketConnection.webSocketFactoryOverride = getMockWebSocketChannel;
+      fakePlatform = MockConnectivityPlatform();
+      ConnectivityPlatform.instance = fakePlatform;
+      final connectivity = Connectivity();
+
+      WebSocketBloc.pollClientOverride = mockHttpClient;
+      WebSocketBloc.connectivityOverride = connectivity;
```
Comment (dnys1) [inline packages/api/amplify_api/test/graphql_test.dart:339]: Can you pass these in the constructor? May be important to have DI when we split out the Dart parts into their own package.

File: packages/api/amplify_api/test/graphql_test.dart [hunk 4]
```diff
@@ -452,6 +449,47 @@ void main() {
...
+        onEstablished: blocReady.complete,
+      );
+
+      await blocReady.future;
+
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+
+      _failClient = true;
+
+      await Future<void>.delayed(const Duration(seconds: 15));
```
Comment (dnys1) [inline packages/api/amplify_api/test/graphql_test.dart:489]: Not sure I follow the 15 seconds - you have a poll interval of 1s and a `_maxFailAttempts = 5` so theoretically only 5s is needed here.

Comment (dnys1) [reply packages/api/amplify_api/test/graphql_test.dart:489]: Can you add a test where `GET /ping` succeeds but returns `400` or a non-`healthy` body.

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart [hunk 5]
```diff
@@ -55,19 +65,41 @@ void main() {
     },
   });
 
+  var failClient = false;
```
Comment (dnys1) [inline packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart:68]: This code is repeated in the `graphql_test` - should we throw it in a method?

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart [hunk 6]
```diff
@@ -117,6 +159,162 @@ void main() {
...
+            isA<ConnectingState>(),
+            isA<ConnectedState>(),
+          ],
+        ),
+      );
+
+      bloc
+          .subscribe(
+        subscribeEvent,
+      )
```
Comment (dnys1) [inline packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart:190]:
```suggestion
      bloc.subscribe(subscribeEvent)
```

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart [hunk 7]
```diff
@@ -117,6 +159,162 @@ void main() {
...
+        bloc.stream,
+        emitsInOrder(
+          [
+            isA<DisconnectedState>(),
+            isA<ConnectingState>(),
+            isA<ConnectedState>(),
+            isA<ReconnectingState>(),
+            isA<ConnectingState>(),
+            isA<ConnectedState>(),
+          ],
```
Comment (dnys1) [inline packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart:224]: Can you add some more clarification to this test? I think here is the best place for it (in the `reason` parameter of `expect`). Basically why we expect only one reconnection when there are 10 toggles below.

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart [hunk 8]
```diff
@@ -117,6 +159,162 @@ void main() {
...
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+      fakePlatform.controller.sink.add(ConnectivityResult.none);
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+      fakePlatform.controller.sink.add(ConnectivityResult.none);
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+      fakePlatform.controller.sink.add(ConnectivityResult.none);
+      fakePlatform.controller.sink.add(ConnectivityResult.wifi);
+    });
+
+    test(
```
Comment (dnys1) [inline packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart:246]: Is this testing something different from the one in `graphql_test`?

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart (hunk 7 above)
Comment (dnys1) [reply packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart:224]: Can you add a test with multiple reconnects, e.g.

```dart
fakePlatform.controller.sink.add(ConnectivityResult.none);

await expectLater(stream, emits(disconnected));

fakePlatform.controller.sink.add(ConnectivityResult.wifi);

await expectLater(stream, emits(connected));

fakePlatform.controller.sink.add(ConnectivityResult.none);

await expectLater(stream, emits(disconnected));

fakePlatform.controller.sink.add(ConnectivityResult.wifi);

await expectLater(stream, emits(connected));
```

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart [hunk 9]
```diff
@@ -117,6 +159,162 @@ void main() {
...
+                isA<ConnectingState>(),
+                isA<ConnectedState>(),
+                isA<ReconnectingState>(),
+                isA<ConnectingState>(),
+                isA<ConnectedState>(),
+              ],
+            ),
+          );
+
+          fakePlatform.controller.sink.add(ConnectivityResult.wifi);
```
Comment (dnys1) [inline packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart:274]: Why is this needed? Connectivity starts with WIFI on right?

**Comment by dnys1** [comment]
Can we expand `graphql_test` to include some assertions about the Stream returned from `API.subscribe` during reconnect? I'm curious what the stream holder should see during a failure, disconnection, and reconnection.

File: packages/api/amplify_api/test/graphql_test.dart (hunk 4 above)
Comment (Equartey) [reply packages/api/amplify_api/test/graphql_test.dart:489]: 15 seconds is tied to the retry strategy duration, not the poll interval. 15 seconds is enough time for 5 exponential back off attempts. I've adjusted the title of the test to reflect this.

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart (hunk 8 above)
Comment (Equartey) [reply packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart]: No, I duplicated here up a layer to ensure functionality in case another rewrite of the bloc happens.

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart [hunk 10]
```diff
@@ -117,6 +141,243 @@ void main() {
...
+            isA<ReconnectingState>(),
+            isA<FailureState>(),
+            isA<PendingDisconnect>(),
+            isA<DisconnectedState>(),
+          ],
+        ),
+      );
+
+      bloc.subscribe(subscribeEvent).listen(
+        (x) {},
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart:368]: nit: Could you replace `(x) {}` w `null`?

**Review by ragingsquirrel3 — COMMENTED** [review]
Still reviewing in more detail, just left some high-level comments and a few nits after pulling down and having a quick look/test.

File: packages/amplify_core/lib/src/types/api/graphql/graphql_subscription_options.dart [hunk 11]
```diff
@@ -15,14 +15,19 @@
...
-/// Configuration options for GraphQL Subscriptions and their WebSockets.
+/// {@template amplify_core.graphql.graphql_subscription_options}
+/// Configuration options for GraphQL subscriptions and their WebSockets.
+/// {@endtemplate}
 class GraphQLSubscriptionOptions {
-  /// Configure the ping interval for AppSync polling for subscription connections.
-  final Duration? pingInterval;
+  /// {@macro amplify_core.graphql.graphql_subscription_options}
+  const GraphQLSubscriptionOptions({
+    this.pollInterval = const Duration(seconds: 30),
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/amplify_core/lib/src/types/api/graphql/graphql_subscription_options.dart:24]: Good call out, yes I moved away from 'ping' phrasing everywhere. Will update commit.

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 12]
```diff
@@ -369,9 +512,29 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
+  Future<AWSBaseHttpResponse> _sendPollRequest() {
+    final req = AWSHttpRequest.get(
+      _currentState.pollUri,
+    );
+
+    return req.send(client: _pollClient).response;
+  }
+
+  /// Sends an exception to all sub blocs and cleans up this bloc
+  Future<void> _shutdownWithException(Exception e) async {
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:535]: I was testing this scenario and noticed that when I hit this shutdown there is no way to establish a new subscription. While I would not expect re-connect to continue once we get here, if user cancels the original subscription, recovers network, tries to create subscription then it never gets established, which I was not expecting.

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 13]
```diff
@@ -252,19 +289,85 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
     logger.verbose('Reset timer');
   }
 
-  Stream<WebSocketState> _shutdown() async* {
-    if (_currentState is ConnectedState) {
-      (_currentState as ConnectedState).timeoutTimer.cancel();
+  /// Handles network loss events by triggering reconnection flow.
```
[the remark this answers is not in the document]
Comment (ragingsquirrel3) [reply packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:292]: can't repro anymore, resolving

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 14]
```diff
@@ -74,10 +85,27 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
 
   late WebSocketState _currentState;
 
+  /// OVERRIDES
+  ///
+  ///
+  ///
+
+  /// The underlying Http client used for poll requests
+  AWSHttpClient get _pollClient => _pollClientOverride ?? AWSHttpClient();
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:103]: Yeah, I hoisted it up to prevent that behavior.

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 15]
```diff
@@ -340,15 +450,48 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
+          break;
+        default:
+          break;
+      }
+    }).listen(null);
+  }
+
+  Future<void> _poll() async {
+    try {
+      final res = await _sendPollRequest();
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:486]: Couldn't hurt having it here. Moved.

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 16]
```diff
@@ -340,15 +450,48 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
     await Future.wait<void>([
-      _subscription.cancel(),
+      _networkSubscription.cancel(),
+      _pollClient.close() as Future<void>,
+      _stateSubscription.cancel(),
       _wsEventController.close(),
       _wsStateController.close(),
     ]);
 
     done.complete();
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:460]: Yes, it's only used in testing. The plugin logic doesn't reuse blocs once they have been closed. Which doesn't happen during reconnection unless it cannot reestablish a connection. We look for done completer during testing to ensure the bloc was cleaned up before moving to the next test.

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 17]
```diff
@@ -221,50 +258,124 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
+    _shutdownWithException(exception);
   }
 
   // Init connection and add channel events to the event stream.
   Stream<WebSocketState> _init(InitEvent event) async* {
     assert(
-      _currentState is DisconnectedState,
-      'Bloc should be in a disconnected state before calling init.',
+      _currentState is! ConnectedState,
+      'Bloc should not be connected state when calling init.',
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:275]: nit: typo "not be in connected ..."

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 18]
```diff
@@ -175,8 +206,8 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
     }
   }
 
-  // Receives connection ack message from ws channel.
-  // Moves ConnectingState => ConnectedState, and registers sub blocs
+  /// Receives connection ack message from ws channel.
+  /// Moves ConnectingState => ConnectedState, and registers sub blocs
```
Comment (ragingsquirrel3) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:210]: Through some wifi on/off toggling in web and unsub/sub I was able to consistently trigger an assertion error from line 214 where the current state is connected. Not sure if this is a blocker bc requires pretty specific order of operations, just a callout.

**Review by ragingsquirrel3 — APPROVED** [review]
🥇 nice work here! I spent a lot of time trying to break this and seeming pretty strong right now, although there are a few edge cases outstanding and surely more to come to light. After merging would be good to rebase the feature branch off next to ensure the integ tests run there, esp w auth. Mentioned offline, but a few followups to keep in mind:
* there is new major version of connectivity_plus so would be good to pull in
* see if can unskip windows integ tests (and fix separately if needed) bc that error related to connectivity_plus

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart [hunk 19]
```diff
@@ -76,90 +99,334 @@ void main() {
...
+            expect(
+              event,
+              isA<ApiException>(),
+            );
+          }),
+        );
+
+        await blocReady.future;
+        mockPollClient.sendUnhealthyResponse = true;
+        await testDone.future;
```
Comment (dnys1) [inline packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart:383]: This shouldn't be needed - expectAsync1 will prevent the test from ending, or should at least.

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 20]
```diff
@@ -47,14 +56,19 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
-    _subscription = blocStream.listen(_emit);
+    _networkSubscription = _getConnectivityStream();
+    _stateSubscription = blocStream.listen(_emit);
   }
 
   @override
   String get runtimeTypeName => 'WebSocketBloc';
 
+  /// Default timeout response for polling
+  final Duration _pollResponseTimeout = const Duration(seconds: 5);
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:70]: We had chosen previously to leave it out to reduce noise on the customer side.

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 21]
```diff
@@ -221,50 +258,124 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
   }
 
-  // Sends registration message on ws channel when connected
-  // else init's connection
+  /// Handle successful polls
+  Stream<WebSocketState> _pollSuccess() async* {
+    final state = _currentState;
+    if (state is ConnectingState) {
+      yield state.reconnecting();
+      add(const ReconnectEvent());
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:321]: Don't think this is needed at all, removed.

File: packages/api/amplify_api/test/util.dart [hunk 22]
```diff
@@ -323,3 +292,40 @@ class MockWebSocketService extends AmplifyWebSocketService {
...
+        return AWSHttpResponse(
+          statusCode: 400,
+          body: utf8.encode('unhealthy'),
+        );
+      }
+      if (induceTimeout) {
+        mockPollFailCount++;
+      }
+
+      if (induceTimeout && mockPollFailCount <= maxFailAttempts) {
```
[the remark this answers is not in the document]
Comment (Equartey) [reply packages/api/amplify_api/test/util.dart:321]: I like it!

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart (hunk 20 above)
[the remark this answers is not in the document]
Comment (dnys1) [reply packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:70]: Can you make it `static const` instead of `final`?

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 23]
```diff
@@ -221,50 +253,131 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
+      add(const InitEvent());
+    } on Exception catch (e) {
+      // Ping failed, close down
+      _shutdownWithException(
+        ApiException(
+          'Unable to recover network connection, web socket will close.',
+          recoverySuggestion: 'Check internet connection.',
+          underlyingException: e,
+        ),
+      );
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:372]:
```suggestion
    } on Exception catch (e, st) {
      // Ping failed, close down
      _shutdownWithException(
        ApiException(
          'Unable to recover network connection, web socket will close.',
          recoverySuggestion: 'Check internet connection.',
          underlyingException: e,
        ),
        st,
      );
```

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 24]
```diff
@@ -221,50 +253,131 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
-  ) async* {
-    yield _currentState.failed();
-
+  Stream<WebSocketState> _connectionError(ConnectionErrorEvent event) async* {
     const exception = ApiException(
       'Error occurred while connecting to the websocket',
     );
 
-    await _sendExceptionToBlocs(exception);
+    yield _currentState.failed(exception);
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:261]: Since the first line of _shutdownWithException emits a failed state we probably don't need this here.

File: packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart [hunk 25]
```diff
@@ -221,50 +253,131 @@ class WebSocketBloc with AWSDebuggable, AmplifyLoggerMixin {
...
+
+    _shutdownWithException(exception);
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/blocs/web_socket_bloc.dart:263]:
```suggestion
    _shutdownWithException(exception, StackTrace.current);
```

File: packages/api/amplify_api/lib/src/graphql/web_socket/state/web_socket_state.dart [hunk 26]
```diff
@@ -114,11 +191,76 @@ class ConnectedState extends WebSocketState {
...
+    _cancelTimers();
+
+    return super.disconnect();
+  }
+
+  /// Move state to [FailureState]
+  @override
+  FailureState failed(
+    Object e, {
+    StackTrace? st,
```
Comment (dnys1) [inline packages/api/amplify_api/lib/src/graphql/web_socket/state/web_socket_state.dart:245]: Can make required. If the stack trace is unavailable (because you've instantiated an exception instead of catching a thrown one), you can use `StackTrace.current`

File: packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart [hunk 27]
```diff
@@ -76,90 +99,333 @@ void main() {
...
+        );
+
+        bloc.subscribe(
+          subscribeEvent,
+        );
+      });
+
+      test('an exception when poll responds unhealthy', () async {
+        final blocReady = Completer<void>();
+        final testDone = Completer<void>();
```
Comment (dnys1) [inline packages/api/amplify_api/test/web_socket/web_socket_bloc_test.dart:348]: Can remove now
